### DISCUSSION TITLE
Recurring/Dynamic spawn ideas

### DISCUSSION THREAD

**Discussion by Tefnut** [cataclysmbn/Cataclysm-BN#3204] [General]
### Summary

Been discussing it in the Discord for a couple weeks now; Wander spawns is very busted and unfun, even with some seeming sanity checks added, but without it, stationary basing rapidly becomes boring, cleared locations rapidly going silent. I've had three ideas that could make recurring monsters spawn, with a focus on them enhancing the experience of a stationary base.

## Random Spawns

- Every interval of time (configurable with both minimum and maximum values to allow variance), monster groups spawn. The amount of spawned groups can also be changed in config with a similar range. 
- The groups pulled from themselves vary, and are weighted, depending on your surroundings; 
- Common surroundings like forests, swamps, rivers, lakes, and fields determine spawns based on a small circle (A 5 overmap tile radius?). 
- Roads and city buildings add zombies to the spawn pool, with certain city buildings (Barricaded houses, pools, and places where children and survivors would be more likely to congregate), weighting relevant special zombies more. 
- While their base influence is lower, meaning a 4-house village will not be overrun by zombies daily, they pull from a wider radius. (10 overmap tiles?)
- Places of interest impact spawns in a wide radius (10 radius for maximum impact, 20 tiles overmap tiles total, maybe), but fall off more the farther off you are. Military facilities, FEMA camps, Shipwrecks, Mi-Go Towers, Fungal structures (maybe), Triffid Groves, and Anthills are some vanilla structures that would influence it; this would be json based so you could have any structure add more.
- Spawns in this setting have 4 requirements; The total spawned in (as in, alive and active) monsters must be below a certain (configurable) threshold, the amount of monsters hostile to the spawn rolled must be below another certain (configurable) threshold, they must not spawn indoors (Though climbing and flying creatures can spawn on roofs), and they must not be in a spot directly visible by the player.

## Triggered Hordes

- This idea takes inspiration from Unturned's Horde Generator. While not a placed generator item, I can imagine some combination of mechanics that would allow periodic spawning of enemy hordes pathing to your location. 
- My current idea is using the Radio and flare gun items; the radio to be able to pick up distant air chatter from the Old Guard, signalling that there might be a plane in the sky that can see you, and the flare gun to signal you have a survivor(s) where you are. While they're not likely lore-wise to assist IIRC, I'm ignoring that for this idea. They could easily be replaced by another altruistic faction.
- Firing the flare into the sky would both start a timer for a small supply airdrop (Which gets larger the more friendly NPCs in the area, and the more flares have been fired in the area), and spawn a horde or, if you're unlucky, a group of bandit NPCs. In either case, they are pathing to the tile the flare was fired from, and the group *also* gets larger the more flares you fire in the area.

## Horde Nights

- This idea is inspired by 7 Days to Die and the flash game Rebuild. In theory, it only applies to stationary bases or sedentary breaks where you stay within a single "Reality bubble" of space
- Every interval (Much like the first option, configurable minimum and maximum) spent in one reality bubble of space, a massive (Size also configurable) horde spawns a distance (also configurable), away from the overmap tile you're in, pathing towards said tile. If you're not there when they arrive, they disperse into random directions, continue the same direction, or just linger around that tile.
- Hordes can be toggled to scale in size with how long the game goes, but to be honest, with natural zombie evolution, things should be fine.
- The interval doesn't reset until you've cleared out either most of a horde, or left that reality bubble and settled down again

These 3 ideas are all theoretical, ideally some form of all 3 would get added to allow players a greater degree of control to if and how their map ages.

**Comment by sigmareaver** [comment]
I hope you don't mind me posting my Discord comment here.

`I feel dynamic spawns is unplayable as long as zombies can spawn inside otherwise secure locations. (e.g. your base). Maybe one easy to implement solution to that would be when picking a random coordinate to spawn a creature in, choose a random edge tile (edge of the map tile) and see if a path exists to the random spawn coordinate. If it does, spawn the creature. If it doesn't, then choose a new random spawn coordinate. Not a perfect solution, but easy to implement.`

I think your ideas sound interesting to play with. Ultimately though, I would still be unable to play with wander spawns as long as they are fundamentally broken. That is, having zombies spawn inside your secure base.

Of course, an exception to that would be a portal spawning monsters, or monsters that can teleport. Likewise, monsters that can fly, jump, or climb could get over walls too. Some tweaks to the creature AI could help too. Wander spawns in its current implementation seems to spawn stuff where the player can't see. Unless that's changed since I last messed with it. But wouldn't it be more interesting if the wander spawns actually wandered into the vicinity? It would require some creature AI improvements I think, in order to give zombies and what not a "direction". Make it so hulks don't path around obstacles either. They just smash through stuff to get to where they want to go and suddenly they're a lot more trouble for a static base.

**Reply by Tefnut** [reply]
All of these would be separate to Wander spawns ideally, turned on separately so we *dont* have to play with the broken system that hasnt been touched practically at all since the DDA days.
